### PR TITLE
Introduce cost log

### PR DESCRIPTION
**Pull request by lukasz-golebiewski**
## Overview
Makes it possible to track which operations incurred what cost

### JIRA ticket:
https://rchain.atlassian.net/browse/RCHAIN-3008

### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

**Comment by lukasz-golebiewski** [comment]
bors try

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/package.scala [hunk 1]
```diff
@@ -1,23 +1,56 @@
...
       _ <- error.ensure(cost.get)(OutOfPhlogistonsError)(_.value >= 0)
+      _ <- writer.tell(Chain.one(amount))
       _ <- cost.modify(_ - amount)
       _ <- error.ensure(cost.get)(OutOfPhlogistonsError)(_.value >= 0)
     } yield ()
 
+  def costLog[M[_]: Sync](): FunctorListen[M, Chain[Cost]] =
+    new DefaultFunctorListen[M, Chain[Cost]] {
+      override val functor: Functor[M]  = implicitly[Functor[M]]
+      private val ref                   = Ref.unsafe(Chain.empty[Cost])
```
Comment (odeac) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/package.scala]: We should avoid using Ref.unsafe. This makes costLog() and all its callers unsafe. 

At this level I would rather keep this abstract function safe by using `Ref.of` instead of `Ref.unsafe` and let the caller decide if he wants to play it safe or not.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/Costs.scala [hunk 2]
```diff
@@ -5,16 +5,20 @@ import com.google.protobuf.ByteString
...
 
 //TODO(mateusz.gorski): Adjust the costs of operations
-final case class Cost(value: Long) extends AnyVal {
-  def *(base: Int): Cost   = Cost(value * base)
-  def +(other: Cost): Cost = Cost(value + other.value)
-  def -(other: Cost): Cost = Cost(value - other.value)
+final case class Cost(value: Long, operation: String) {
+  def *(base: Int): Cost   = Cost(value * base, s"$operation * $base")
+  def +(other: Cost): Cost = Cost(value + other.value, "")
+  def -(other: Cost): Cost = Cost(value - other.value, "")
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/Costs.scala]: discarding the other's 'operation' seems wrong in here

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/Costs.scala [hunk 3]
```diff
@@ -5,16 +5,20 @@ import com.google.protobuf.ByteString
 import coop.rchain.models.{PCost, Par, ProtoM, StacksafeMessage}
 
 //TODO(mateusz.gorski): Adjust the costs of operations
-final case class Cost(value: Long) extends AnyVal {
-  def *(base: Int): Cost   = Cost(value * base)
-  def +(other: Cost): Cost = Cost(value + other.value)
-  def -(other: Cost): Cost = Cost(value - other.value)
+final case class Cost(value: Long, operation: String) {
+  def *(base: Int): Cost   = Cost(value * base, s"$operation * $base")
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/Costs.scala]: please add parentheses, especially after including the discarded 'operation'-s for + and -

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/package.scala [hunk 4]
```diff
@@ -1,23 +1,56 @@
...
-  )(implicit cost: _cost[F], error: _error[F]): F[Unit] =
+  )(implicit cost: _cost[F], error: _error[F], writer: _costWriter[F]): F[Unit] =
     for {
       _ <- error.ensure(cost.get)(OutOfPhlogistonsError)(_.value >= 0)
+      _ <- writer.tell(Chain.one(amount))
       _ <- cost.modify(_ - amount)
       _ <- error.ensure(cost.get)(OutOfPhlogistonsError)(_.value >= 0)
     } yield ()
 
+  def costWriter[M[_]: Sync](): FunctorListen[M, Chain[Cost]] =
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/package.scala]: I think 'FunctorTell' should suffice - we should only ever need the cost value at the end of the world.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/matcher/MatchTest.scala [hunk 5]
```diff
@@ -935,29 +937,49 @@ class VarMatcherSpec extends FlatSpec with Matchers with TimeLimits with TripleE
...
     res should be(Left(OutOfPhlogistonsError))
   }
 
   it should "charge for matching operations" in {
-    val (phloLeft, _) = doMatchAndCharge(initialPhlo = 100)
+    val ((phloLeft, _), log) = doMatchAndCharge(initialPhlo = 100)
     phloLeft.value shouldBe 90
+    log.toList should contain theSameElementsAs List(
+      Cost(4, "equality check"),
+      Cost(0, "comparison * 0"),
```
Comment (ArturGajowy) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/matcher/MatchTest.scala]: we probably shouldn't ever allow a `Cost(operation = "")`. This way, this line would state e.g. 
`Cost(0, "comparison * number of patterns")`
or
`Cost(0, "comparison * number of sends")`

Comment (ArturGajowy) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/matcher/MatchTest.scala]: The zeroes are worrying, but that's a different story

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala [hunk 6]
```diff
@@ -19,7 +19,7 @@ trait Substitute[M[_], A] {
 }
 
 object Substitute {
-  private[interpreter] def charge[A: Chargeable, M[_]: Sync: _cost: _error](
+  private[interpreter] def charge[A: Chargeable, M[_]: Sync: _cost: _costWriter: _error](
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: We must not thread a separate `_costWriter` capability. Instead, we should redefine `_cost` to include this capability.

Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: The way this should work, is:

we'd have a standard `_cost` instance that does no logging
we'd have a `_cost` instance that does the logging, and requires `FunctorTell[F, Cost]` for **creating** (not using!) it.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/package.scala [hunk 7]
```diff
@@ -1,23 +1,56 @@
...
+
+  def noOpCostWriter[M[_]: Applicative]: _costWriter[M] =
+    new DefaultFunctorTell[M, Chain[Cost]] {
+      override val functor: Functor[M]  = implicitly[Functor[M]]
+      def tell(l: Chain[Cost]): M[Unit] = Applicative[M].pure(())
+    }
+
+  implicit def ntCostWriter[F[_]: Monad: _costWriter, G[_]: Functor](
+      nt: F ~> G
+  ): _costWriter[G] =
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/package.scala]: Why is this needed? I think it might no logner be needed after we make `_cost` do the writing transparently.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 8]
```diff
@@ -67,7 +67,7 @@ object Reduce {
       s: Sync[M],
       fTell: FunctorTell[M, Throwable],
       cost: _cost[M],
-      costWriter: _costWriter[M],
+      costLog: _costLog[M],
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: From now on, before submitting a PR, please squash all the `rename` commits in a PR with the commit introducing the thing being renamed.

Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: Please fix this for this PR as well.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/Costs.scala [hunk 9]
```diff
@@ -39,12 +40,12 @@ trait Costs {
       "equality check"
     )
 
-  final val BOOLEAN_AND_COST = Cost(2)
-  final val BOOLEAN_OR_COST  = Cost(2)
+  final val BOOLEAN_AND_COST = Cost(2, "boolean and")
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/Costs.scala]: This should be part of the first commit.

Alternatively, a commit log allowing separation of those concerns would be:

```
Add `operation` field to cost (remove the ability to pass no/empty operation name, combine operation names for * - +)

Add a logging variant of the `_cost` effect (uses FunctorTell[F, Chain[Cost]]) underneath
```

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/Costs.scala (hunk 2 above)
Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/Costs.scala]: Please include the `other` oepration name for + and -

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/Costs.scala (hunk 3 above)
Comment (lukasz-golebiewski) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/Costs.scala]: Where?

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala (hunk 6 above)
Comment (lukasz-golebiewski) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: How do you want to "include" `FunctorTell` in a `MonadState`?

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala (hunk 8 above)
Comment (lukasz-golebiewski) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: Ok, will do

**Comment by lukasz-golebiewski** [comment]
bors try

**Comment by lukasz-golebiewski** [comment]
bors try

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/package.scala (hunk 1 above)
Comment (lukasz-golebiewski) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/package.scala]: This implementation is currently used only in tests. I moved it there

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/package.scala [hunk 10]
```diff
@@ -25,6 +25,11 @@ package object matcher {
 
   type MatcherMonadT[F[_], A] = StateT[StreamT[F, ?], FreeMap, A]
 
+  import coop.rchain.rholang.interpreter.matcher.StreamT
+
+  implicit def matcherMonadCostLog[F[_]: Monad: _cost](): _cost[MatcherMonadT[F, ?]] =
+    ntCostLog(λ[F ~> MatcherMonadT[F, ?]](fa => StateT.liftF(StreamT.liftF(fa))))
```
[the remark this answers is not in the document]
Comment (lukasz-golebiewski) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/package.scala]: This is required to derive a `_cost` instance for the `MatcherMonad` used by the `SpatialMatcher`. The reason could be the `StreamT`, or better said, the lack of some required derivation rule. I guess this could be investigated and improved in a separate PR

**Comment by lukasz-golebiewski** [comment]
bors cancel

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/Costs.scala (hunk 2 above)
Comment (lukasz-golebiewski) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/Costs.scala]: This change will have to be reverted, otherwise the heap will run out of space
